File: ts/src/watcher/dedup.ts
/**
 * Time-windowed deduplication cache keyed on `(txHash, logIndex)`.
 * TTL is measured from insertion time; reads do not extend the TTL.
 */
export class DedupCache {
  private cache = new Map<string, number>();
  private ttlMs: number;
  private maxSize: number;

  constructor(ttlMs = 60_000, maxSize = 10_000) {
    this.ttlMs = ttlMs;
    this.maxSize = maxSize;
  }

  has(txHash: string, logIndex: number): boolean {
    const key = `${txHash.toLowerCase()}:${logIndex}`;
    const ts = this.cache.get(key);
    if (ts === undefined) return false;
    if (Date.now() - ts > this.ttlMs) {
      this.cache.delete(key);
      return false;
    }
    return true;
  }

  add(txHash: string, logIndex: number): void {
    const key = `${txHash.toLowerCase()}:${logIndex}`;
    this.cache.set(key, Date.now());
    this.evict();
  }

  private evict(): void {
    if (this.cache.size <= this.maxSize) return;
    const now = Date.now();
    for (const [key, ts] of this.cache) {
      if (now - ts > this.ttlMs) {
        this.cache.delete(key);
      }
    }

    if (this.cache.size > this.maxSize) {
      const entries = [...this.cache.entries()];
      entries.sort((a, b) => a[1] - b[1]);
      const toRemove = entries.slice(0, entries.length - this.maxSize);
      for (const [key] of toRemove) {
        this.cache.delete(key);
      }
    }
  }

  clear(): void {
    this.cache.clear();
  }

  get size(): number {
    return this.cache.size;
  }
}

File: rs/tempo-reconcile/src/watcher/dedup.rs
use std::collections::HashMap;
use std::time::{Duration, Instant};

/// TTL-based deduplication cache keyed by `"{tx_hash}:{log_index}"`.
///
/// Eviction is lazy: expired entries are removed only when the cache reaches `max_size`.
pub struct DedupCache {
    inner: HashMap<String, Instant>,
    ttl: Duration,
    max_size: usize,
}

impl DedupCache {
    pub fn new(ttl_secs: u64, max_size: usize) -> Self {
        Self {
            inner: HashMap::new(),
            ttl: Duration::from_secs(ttl_secs),
            max_size: max_size.max(1),
        }
    }

    /// Returns `true` and marks `key` as seen if it has not been seen within the TTL window.
    /// Returns `false` (duplicate) otherwise.
    pub fn check_and_insert(&mut self, key: &str) -> bool {
        let now = Instant::now();

        if let Some(ts) = self.inner.get(key) {
            if now.duration_since(*ts) < self.ttl {
                return false;
            }
        }

        // Key is new or expired. Evict to make room if at capacity.
        if self.inner.len() >= self.max_size {
            // Phase 1: evict expired entries
            self.inner
                .retain(|_, ts| now.duration_since(*ts) < self.ttl);

            // Phase 2: if still at capacity, evict oldest (LRU)
            if self.inner.len() >= self.max_size {
                if let Some(oldest_key) = self
                    .inner
                    .iter()
                    .min_by_key(|(_, ts)| *ts)
                    .map(|(k, _)| k.clone())
                {
                    self.inner.remove(&oldest_key);
                }
            }
        }

        self.inner.insert(key.to_string(), now);
        true
    }
}
